### DISCUSSION TITLE
hidden pitfalls:  the use of all-zero codewords as valid test patterns

### DISCUSSION THREAD

**Discussion by WuShengHan** [NVlabs/sionna#641] [Q&A]
I’d like to discuss an issue in ECC verification, specifically regarding the use of all-zero or all-one codewords as valid test patterns. Using these can create misleading results, particularly in quantized LLR (Log-Likelihood Ratio) calculations, affecting how we perceive the error correction capability.

Here’s the concern:

If all-zero is treated as a valid codeword, simulating by flipping one bit from all-zero doesn’t accurately reflect typical conditions. Ideally, LLR values of zero should indicate equal likelihood for bits being 0 or 1. However, with all-zero input, LLR=0 leads to a bias where bits are consistently interpreted as 0, giving an overly optimistic view of the ECC’s correction ability.

In real applications, data often contains a roughly balanced mix of 1s and 0s. This means LLR=0 shouldn’t predictably resolve the bit’s value, as it does in all-zero cases. Consequently, actual error correction performance could see a significant gap, leading to reduced performance when deployed in real-world scenarios.

This discrepancy is particularly evident in the error floor and large-scale data validation stages, resulting in bugs that become problematic only in final stages. Such issues compromise product performance, potentially causing next-generation transmission quality to fall short—akin to some of the challenges encountered in 5G.

To address this, I propose a more rigorous verification process, aiming to avoid discrepancies between testing and real-world application.

**Comment by SebastianCa** [comment]
Hi @WuShengHan,

You’re correct; decoding results can indeed be biased in all-zero codeword simulations. For this reason, Sionna provides an [LDPC 5G encoder](https://nvlabs.github.io/sionna/api/fec.ldpc.html#sionna.fec.ldpc.encoding.LDPC5GEncoder) as well as an [encoder for linear codes](https://nvlabs.github.io/sionna/api/fec.linear.html#sionna.fec.linear.LinearEncoder).

**Comment by WuShengHan** [comment]
Do you consider that  add a warning message in the all-zero encoder function on GitHub to explain the misleading simulation. 
This could clarify why it may be better to replot using a codeword with a mix of 0s and 1s, avoiding any misleading results.
For example, 5G LDPC and Polar codes often have high error floors, and generating a BER of 10^-6 (one failure per 100-mega-byte of data stream ) only takes a minute or two with CPU AVX, suggesting the platform could be better optimized to showcase GPU acceleration capabilities.

The repo may help many researchers to know error floor issues of channel coding that 6G need to solve.
